Fix the FORM package

tueda opened the pull request:
- The first commit aims to fetch the correct tarballs.
- The second one fixes the dependency.

---

The first one may need some explanation. Currently, `form` executables installed via Spack do not know revision information correctly. Indeed, the build dates are printed:
```
# form@4.2.1
$ form -v
FORM 4.2 (Oct  2 2021) 64-bits
  0.00 sec out of 0.00 sec
```
```
# form@4.1-20131025
$ form -v
FORM 4.1 (Oct  2 2021) 64-bits
  0.00 sec out of 0.00 sec
```
This is because the fetched tarballs are not intended to be used for build and do not contain detailed revision information. You see some warnings about the revision information on running `spack install -v form@4.2.1`:
```
========================================================================
Failed to determine the revision of the source code.

The reason may be
  - this is neither a source distribution (containing the configure
    script) nor a cloned Git repository,
  - this is a shallow clone and no version tags are reachable,
  - some required utilities (e.g., git) are missing.

Source distributions and some binaries can be found in:

  http://www.nikhef.nl/~form/maindir/binaries/binaries.html
  https://github.com/vermaseren/form/releases

The latest source code can be cloned by:

  git clone https://github.com/vermaseren/form.git

You can continue the build, but binaries will not contain the revision
information.
========================================================================
```
After applying the first commit, `form` prints the revision information correctly:
```
# form@4.2.1
$ form -v
FORM 4.2.1 (Nov 21 2018, v4.2.1) 64-bits
  0.00 sec out of 0.00 sec
```
```
# form@4.1-20131025
$ form -v
FORM 4.1 (Oct 25 2013) 64-bits
  0.00 sec out of 0.00 sec
```

tueda committed:
form: fetch the correct tarballs

* This fixes the revision information printed by "form -v".

* Remove unused build dependency, we don't need autoreconf.

Co-authored-by: bernhardkaindl <43588962+bernhardkaindl@users.noreply.github.com>

Co-authored-by: iarspider <iarspider@gmail.com>

tueda committed:
form: fix dependent libraries (gmp <-> zlib)

iarspider commented on var/spack/repos/builtin/packages/form/package.py:
```diff
@@ -10,18 +10,18 @@ class Form(AutotoolsPackage):
-    version('4.2.1', sha256='6f32c7470d00e8ab6934dc352f5a78e29290146a00e5775f8cd5fef7810bbbb8')
-    version('4.1-20131025', sha256='caece2c6e605ccf32eb3612c4ed5c9257a7a62824ad219c5e46b6d00066f1ba6')
+    version('4.2.1', sha256='f2722d6d4ccb034e01cf786d55342e1c21ff55b182a4825adf05d50702ab1a28')
+    version('4.1-20131025', sha256='fb3470937d66ed5cb1af896b15058836d2c805d767adac1b9073ed2df731cbe9')
```
```suggestion
    version('4.1-20131025', sha256='fb3470937d66ed5cb1af896b15058836d2c805d767adac1b9073ed2df731cbe9',
            url='https://github.com/vermaseren/form/releases/download/v4.1-20131025/form-4.1.tar.gz')
```

and remove `url_for_version` function

tueda replied:
Thank you for your review. I will make a change following your suggestion.

iarspider replied:
You don't need `url_for_version` function if you pass `url` parameter

bernhardkaindl replied:
AFAICS, this is pushed now as commmit 67165dd by Takahiro

iarspider replied:
Yup. Still getting used to reviewing, sorry!

bernhardkaindl commented on var/spack/repos/builtin/packages/form/package.py:
```diff
@@ -10,18 +10,18 @@ class Form(AutotoolsPackage):
+    version('4.2.1', sha256='f2722d6d4ccb034e01cf786d55342e1c21ff55b182a4825adf05d50702ab1a28')
+    version('4.1-20131025', sha256='fb3470937d66ed5cb1af896b15058836d2c805d767adac1b9073ed2df731cbe9')
 
     depends_on('autoconf', type='build')
     depends_on('automake', type='build')
     depends_on('libtool',  type='build')
     depends_on('m4',       type='build')
```
```suggestion
```
Since you switched to release tarballs which contain a `configure` script, `AutotoolsPackage` no longer needs to generate it using `autoreconf`, so the autotools dependencies are obsolete (including m4), tested it. You can just accept the suggestion using the button below. With that, added, you can request a review from me.

tueda replied:
Thank you for the review. I'm fine with this.

bernhardkaindl reviewed (changes requested):
Please apply my suggestion to remove the 4 autoreconf dependencies, tested it and I was co-authoring the last updated of the related code in the AutotoolsPackage: t is ok to remove these for the release tarballs (not used for them).

tueda commented:
Another point: can we switch `gmp` and `zlib` variants to `default=True`? (I'm not sure whether there is some Spack community policy like default variants should be minimal.) As written in the [FORM Wiki](https://github.com/vermaseren/form/wiki/Installation#building-from-the-source-distribution), they are recommended to use:
> We recommend building FORM with two external libraries, GMP and zlib.

iarspider commented:
Sorry, I thought my review was visible. One minor change, otherwise LGTM

bernhardkaindl commented:
> Another point: can we switch `gmp` and `zlib` variants to `default=True`? (I'm not sure whether there is some Spack community policy like default variants should be minimal.) As written in the [FORM Wiki](https://github.com/vermaseren/form/wiki/Installation#building-from-the-source-distribution), they are recommended to use:
> 
> > We recommend building FORM with two external libraries, GMP and zlib.

Good reference! :+1:  By all means, and because of gmp and zlib are not heavvy dependencies, it would be ok for me.

But I'm not the maintainer, I don't use this package, so I'm referring this to @iarspider for his take!

@tueda, from the feel of it, you seem to care about the package, so if ok for @iarspider, would be be content to add yourself as a 2nd maintainer after iarspider? You'd only be notified by spackbot like iarspider was when somebody submits a PR, there is no mandatory response needed on your part, just people would know that you can be asked if there is a question or need of a review.

@iarspider, what do you think, is it worth it enable those by default (and maybe if ok for him, have him as 2nd maintainer) ?

bernhardkaindl reviewed (dismissed):
If you like to be finished, you can just push enabing gmp and zlib by default and wait for iarspider's opinion/review on it.

tueda committed:
form: enable gmp and zlib by default

bernhardkaindl reviewed (dismissed):
Since I saw that iarspider approved this commit, let's merge it! :100:

tueda commented:
Rebased with the removal of `url_for_version`. (I hope it works with your co-authorship.)
I also added the 3rd commit to enable `gmp` and `zlib` by default.

If you guys are OK, I am fine to become the 2nd maintainer.

iarspider commented:
I'm fine with both changes.

bernhardkaindl commented on var/spack/repos/builtin/packages/form/package.py:
```diff
@@ -10,22 +10,19 @@ class Form(AutotoolsPackage):
     homepage = "https://www.nikhef.nl/~form/"
-    url      = "https://github.com/vermaseren/form/archive/v4.2.1.tar.gz"
+    url      = "https://github.com/vermaseren/form/releases/download/v4.2.1/form-4.2.1.tar.gz"
     maintainers = ['iarspider']
```
```suggestion
    maintainers = ['iarspider', 'tueda']
```
@iarspider I guess this also meant your are fine with this too? (you can commit my suggestion with the button above)

tueda replied:
Sure.

iarspider replied:
Yes - both adding tueda as a maintainer and changing gmp and zlib defaults

bernhardkaindl replied:
Done, that's all from me as well! :+1: Thanks!

tueda replied:
Oh, sorry, I pushed the button without reading your text carefully...

tueda committed:
form: add a 2nd maintainer

Co-authored-by: Bernhard Kaindl <43588962+bernhardkaindl@users.noreply.github.com>

bernhardkaindl commented:
Done! :1st_place_medal: :-)

In case the check `ci/gitlab-ci`, fails that's not fatal! (it has problems not related to this package and is being worked on).
We can merge without `ci/gitlab-ci` at the moment.

In case you wonder about the commit message:
For the upcoming auto-squash, I updated it try to follow the usual commit message style for spack with all commit comments below that squashed to just the Co-autored lines.

**Thanks for the contribution, to be fine to be listed as 2nd maintainer and thanks for the review with suggestion iarspider!**

bernhardkaindl commented:
Looks good, I've built the final merged commit from develop, *I think that's it for this one!* :+1:
